FSW/src/util/hammingBlock.cpp

```diff
@@ -0,0 +1,269 @@
+            
+            // if message bit is 1, include in calculation of indexParity
+            if (checkBit(message, messageIdx)) {
+                indexParity ^= hammingIdx;
+                totalSetBits++;
+            }
+            messageIdx++;
+        }
+    }
+    
+    /* Set parity bits in hamming block 
+    *   copies each bit of index parity to the hamming block at the index equal to its value.
+    *   e.g. for an index parity of 0001 0011, parity bits will be set at indices 1,2,16 in the hamming block */
+    for (int n = 0; n < 0b1000 - 1; n++) { // for each bit in indexParity...
+        int idx = pow(2, n); // find corresponding index in hamming block
+        assignBit(m_block, idx, BIT_CHECK(indexParity, n)); // set parity bit in hamming block
+    }
+    
+    // find total block parity and put it in index 0
+    bool blockParity = !!(totalSetBits % 2);
+    assignBit(m_block, 0, blockParity);
+} 
+
+/* - - - - - - getMessage - - - - - - *
+ * Usage:
+ *  Retrieves and returns the decoded message
+ * 
+ * Inputs:
+ *  None
+ * 
+ * Outputs:
+ *  pointer to the decoded message, type uint8_t*
+ */
+uint8_t *HammingBlock::getMessage() {
+    memset(m_message, 0, MESSAGE_SIZE); // clear last known message
+    // Extract message from block
+    int messageIdx = 0;
+    for (int hammingIdx = 0; hammingIdx < HAMMING_BLOCK_SIZE * 8; hammingIdx++) { // for each bit in the hamming block...
+        if (( hammingIdx & (hammingIdx - 1) ) != 0) { // if current index is not power of 2, and thus not a parity bit...
+            // copy bit from hamming block to destination message block
+            assignBit(m_message, messageIdx, checkBit(m_block, hammingIdx));
+            messageIdx++;
+        }
+    }
+    return m_message;
+}
+
+/* - - - - - - scanBlock - - - - - - *
+ * Usage:
+ *  Scans the block for errors.
+ * 
+ * Inputs:
+ *  None
+ *  
+ * Outputs:
+ *  ErrorReport struct with size of detected error and error position. Returns a position of -1 for errors not of size 1
+ */
+ErrorReport HammingBlock::scanBlock() {
+    ErrorReport errorInfo;
+    uint8_t indexParity = 0;
+    int totalSetBits = 0;
+
+    // calculate index parity
+    for (int hammingIdx = 0; hammingIdx < HAMMING_BLOCK_SIZE * 0b1000; hammingIdx++) { // for each bit in the hamming block...
+        if (checkBit(m_block, hammingIdx)) {
+                indexParity ^= hammingIdx;
+                totalSetBits++;
+        }
+    }
+    // calculate block parity
+    bool blockParity = !!(totalSetBits % 2);
+    
+    // determine error type
+    if (checkBit(m_block, 0) == blockParity) {
+        if (indexParity == 0) {
+            // no error detected
+        } else {
+            // two bit error detected
+            errorInfo.size = 2;
+        }
+    } else {
+        // single bit error detected
+        errorInfo.size = 1;
+        errorInfo.position = indexParity; // The parity tells us the index of the error!!!
```

Comment:
Comments are generally doing a solid job explaining what's going on here, some more in detail documentation will definitely be needed here though if we want to make this accessible to next year's team.

Reply:
Oh yes, some serious documentation is called for to make this understandable to future teams, especially since they will need to modify it to add pointing data. You should have seen it before I encapsulated this in classes — it worked, but trying to understand it was hopeless